File: packages/flutter/lib/src/url.dart
import 'dart:convert';

bool isInlineSvg(String source) {
  final t = source.trimLeft();
  return t.startsWith('<svg') || t.startsWith('<?xml');
}

String? decodeDataUrl(String source) {
  final match = RegExp(
    r'^data:image/svg\+xml(?:;charset=[^;,]+)?(;base64)?,(.*)$',
    caseSensitive: false,
  ).firstMatch(source);
  if (match == null) return null;
  final isBase64 = match.group(1) != null;
  final data = match.group(2) ?? '';
  try {
    if (isBase64) {
      return String.fromCharCodes(base64Decode(data));
    }
    return Uri.decodeComponent(data);
  } catch (_) {
    return null;
  }
}

bool isSafeUrl(String value) {
  final v = value.trim().replaceAll(RegExp(r"^['\"]|['\"]$"), '');
  if (v.isEmpty || v.startsWith('#')) return true;
  if (v.startsWith('//')) return true;
  final schemeMatch = RegExp(r'^([a-zA-Z][a-zA-Z0-9+.\-]*):').firstMatch(v);
  if (schemeMatch == null) return true;
  final scheme = schemeMatch.group(1)?.toLowerCase() ?? '';
  if (scheme == 'http' || scheme == 'https' || scheme == 'blob') return true;
  if (scheme == 'data') {
    return RegExp(r'^data:image/', caseSensitive: false).hasMatch(v);
  }
  return false;
}

bool hasUnsafeUrl(String value) {
  final pattern = RegExp(r'url\(([^)]+)\)', caseSensitive: false);
  for (final m in pattern.allMatches(value)) {
    final raw = (m.group(1) ?? '').trim();
    if (!isSafeUrl(raw)) return true;
  }
  return false;
}

File: packages/svg/src/core/url.ts
export const isInlineSvg = (source: string) => {
  const trimmed = source.trim();
  return trimmed.startsWith("<svg") || trimmed.startsWith("<?xml");
};

export const decodeDataUrl = (source: string) => {
  const match = source.match(/^data:image\/svg\+xml(?:;charset=[^;,]+)?(;base64)?,(.*)$/i);
  if (!match) return null;
  const isBase64 = Boolean(match[1]);
  const data = match[2] ?? "";
  try {
    if (isBase64) {
      if (typeof atob === "function") return atob(data);
      return null;
    }
    return decodeURIComponent(data);
  } catch {
    return null;
  }
};

export const isSafeUrl = (value: string) => {
  const trimmed = value.trim().replace(/^['"]|['"]$/g, "");
  if (!trimmed || trimmed.startsWith("#")) return true;
  if (trimmed.startsWith("//")) return true;
  const schemeMatch = trimmed.match(/^([a-zA-Z][a-zA-Z0-9+.-]*):/);
  if (!schemeMatch) return true;
  const scheme = schemeMatch[1];
  if (!scheme) return false;
  const normalized = scheme.toLowerCase();
  if (normalized === "http" || normalized === "https" || normalized === "blob") {
    return true;
  }
  if (normalized === "data") return /^data:image\//i.test(trimmed);
  return false;
};

export const hasUnsafeUrl = (value: string) => {
  const pattern = /url\(([^)]+)\)/gi;
  let match: RegExpExecArray | null;
  while ((match = pattern.exec(value))) {
    const raw = match[1]?.trim() ?? "";
    if (!isSafeUrl(raw)) return true;
  }
  return false;
};
